File: docs/Greeter.js
/*
 * LightDMGreeter.js
 *
 * Copyright © 2017 Antergos Developers <dev@antergos.com>
 *
 * This file is part of Web Greeter.
 *
 * Web Greeter is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Web Greeter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * The following additional terms are in effect as per Section 7 of the license:
 *
 * The preservation of all legal notices and author attributions in
 * the material or in the Appropriate Legal Notices displayed
 * by works containing it is required.
 *
 * You should have received a copy of the GNU General Public License
 * along with web-greeter; If not, see <http://www.gnu.org/licenses/>.
 */


/**
 * Base class for the greeter's Theme JavaScript API. Greeter themes will interact
 * directly with an object derived from this class to facilitate the user log-in process.
 * The greeter will automatically create an instance when it starts.
 * The instance can be accessed using the global variable: `lightdm`.
 *
 * @memberOf LightDM
 */
class Greeter {

	constructor() {
		if ( 'lightdm' in window ) {
			return window.lightdm;
		}

		window.lightdm = ThemeUtils.bind_this( this );

		return window.lightdm;
	}

	/**
	 * The username of the user being authenticated or {@link null}
	 * if no authentication is in progress
	 * @type {String|Null}
	 * @readonly
	 */
	get authentication_user() {}

	/**
	 * Whether or not the guest account should be automatically logged
	 * into when the timer expires.
	 * @type {Boolean}
	 * @readonly
	 */
	get autologin_guest() {}

	/**
	 * The number of seconds to wait before automatically logging in.
	 * @type {Number}
	 * @readonly
	 */
	get autologin_timeout() {}

	/**
	 * The username with which to automattically log in when the timer expires.
	 * @type {String}
	 * @readonly
	 */
	get autologin_user() {}

	/**
	 * The battery data
	 * @type {Battery}
	 * @readonly
	 */
	get batteryData() {}

	/**
	 * The display brightness
	 * @type {Number}
	 */
	get brightness() {}
	set brightness( quantity ) {}

	/**
	 * Whether or not the greeter can access to battery data.
	 * @type {boolean}
	 * @readonly
	 */
	get can_access_battery() {}

	/**
	 * Whether or not the greeter can control display brightness.
	 * @type {boolean}
	 * @readonly
	 */
	get can_access_brightness() {}

	/**
	 * Whether or not the greeter can make the system hibernate.
	 * @type {Boolean}
	 * @readonly
	 */
	get can_hibernate() {}

	/**
	 * Whether or not the greeter can make the system restart.
	 * @type {Boolean}
	 * @readonly
	 */
	get can_restart() {}

	/**
	 * Whether or not the greeter can make the system shutdown.
	 * @type {Boolean}
	 * @readonly
	 */
	get can_shutdown() {}

	/**
	 * Whether or not the greeter can make the system suspend/sleep.
	 * @type {Boolean}
	 * @readonly
	 */
	get can_suspend() {}

	/**
	 * The name of the default session.
	 * @type {String}
	 * @readonly
	 */
	get default_session() {}

	/**
	 * Whether or not guest sessions are supported.
	 * @type {Boolean}
	 * @readonly
	 */
	get has_guest_account() {}

	/**
	 * Whether or not user accounts should be hidden.
	 * @type {boolean}
	 * @readonly
	 */
	get hide_users_hint() {}

	/**
	 * The system's hostname.
	 * @type {String}
	 * @readonly
	 */
	get hostname() {}

	/**
	 * Whether or not the greeter is in the process of authenticating.
	 * @type {Boolean}
	 * @readonly
	 */
	get in_authentication() {}

	/**
	 * Whether or not the greeter has successfully authenticated.
	 * @type {Boolean}
	 * @readonly
	 */
	get is_authenticated() {}

	/**
	 * The current language or {@link null} if no language.
	 * @type {Language|Null}
	 * @readonly
	 */
	get language() {}

	/**
	 * A list of languages to present to the user.
	 * @type {Language[]}
	 * @readonly
	 */
	get languages() {}

	/**
	 * The currently active layout for the selected user.
	 * @type {Layout}
	 */
	get layout() {}
	set layout(layout) {}

	/**
	 * A list of keyboard layouts to present to the user.
	 * @type {Layout[]}
	 * @readonly
	 */
	get layouts() {}

	/**
	 * Whether or not the greeter was started as a lock screen.
	 * @type {Boolean}
	 * @readonly
	 */
	get lock_hint() {}

	/**
	 * Whether or not the guest account should be selected by default.
	 * @type {Boolean}
	 * @readonly
	 */
	get select_guest_hint() {}

	/**
	 * The username to select by default.
	 * @type {String}
	 * @readonly
	 */
	get select_user_hint() {}

	/**
	 * List of available sessions.
	 * @type {Session[]}
	 * @readonly
	 */
	get sessions() {}

	/**
	 * Check if a manual login option should be shown. If {@link true}, the theme should
	 * provide a way for a username to be entered manually. Otherwise, themes that show
	 * a user list may limit logins to only those users.
	 * @type {Boolean}
	 * @readonly
	 */
	get show_manual_login_hint() {}

	/**
	 * Check if a remote login option should be shown. If {@link true}, the theme should provide
	 * a way for a user to log into a remote desktop server.
	 * @type {Boolean}
	 * @readonly
	 * @internal
	 */
	get show_remote_login_hint() {}

	/**
	 * List of available users.
	 * @type {User[]}
	 * @readonly
	 */
	get users() {}


	/**
	 * Starts the authentication procedure for a user.
	 *
	 * @param {String|Null} username A username or {@link null} to prompt for a username.
	 */
	authenticate( username ) {}

	/**
	 * Starts the authentication procedure for the guest user.
	 */
	authenticate_as_guest() {}

	/**
	 * Updates the battery data
	 */
	batteryUpdate() {}

	/**
	 * Set the brightness to quantity
	 * @param {Number} quantity The quantity to set
	 */
	brightnessSet( quantity ) {}

	/**
	 * Increase the brightness by quantity
	 * @param {Number} quantity The quantity to increase
	 */
	brightnessIncrease( quantity ) {}

	/**
	 * Decrease the brightness by quantity
	 * @param {Number} quantity The quantity to decrease
	 */
	brightnessDecrease( quantity ) {}

	/**
	 * Cancel user authentication that is currently in progress.
	 */
	cancel_authentication() {}

	/**
	 * Cancel the automatic login.
	 */
	cancel_autologin() {}

	/**
	 * Triggers the system to hibernate.
	 * @returns {Boolean} {@link true} if hibernation initiated, otherwise {@link false}
	 */
	hibernate() {}

	/**
	 * Provide a response to a prompt.
	 * @param {*} response
	 */
	respond( response ) {}

	/**
	 * Triggers the system to restart.
	 * @returns {Boolean} {@link true} if restart initiated, otherwise {@link false}
	 */
	restart() {}

	/**
	 * Set the language for the currently authenticated user.
	 * @param {String} language The language in the form of a locale specification (e.g.
	 *     'de_DE.UTF-8')
	 * @returns {Boolean} {@link true} if successful, otherwise {@link false}
	 */
	set_language( language ) {}

	/**
	 * Triggers the system to shutdown.
	 * @returns {Boolean} {@link true} if shutdown initiated, otherwise {@link false}
	 */
	shutdown() {}

	/**
	 * Start a session for the authenticated user.
	 * @param {String|null} session The session to log into or {@link null} to use the default.
	 * @returns {Boolean} {@link true} if successful, otherwise {@link false}
	 */
	start_session( session ) {}

	/**
	 * Triggers the system to suspend/sleep.
	 * @returns {Boolean} {@link true} if suspend/sleep initiated, otherwise {@link false}
	 */
	suspend() {}

	/**
	 * Gets emitted when the greeter has completed authentication.
	 * @type {Signal}
	 */
	authentication_complete;

	/**
	 * Gets emitted when the automatic login timer has expired.
	 * @type {Signal}
	 */
	autologin_timer_expired;

	/**
	 * Gets emitted when brightness is updated
	 * @type {Signal}
	 */
	brightness_update;

	/**
	 * Gets emitted when the user has logged in and the greeter is no longer needed.
	 * @type {Signal}
	 */
	idle;

	/**
	 * Gets emitted when the user is returning to a greeter that
	 * was previously marked idle.
	 * @type {Signal}
	 */
	reset;

	/**
	 * Gets emitted when the greeter should show a message to the user.
	 * @type {Signal}
	 */
	show_message;

	/**
	 * Gets emitted when the greeter should show a prompt to the user.
	 * @type {Signal}
	 */
	show_prompt;

}

/**
 * JS-Cookie instance - Themes must manually load the included vendor script in order to use this object.
 * @name Cookies
 * @type {object}
 * @version 2.1.3
 * @memberOf window
 * @see [JS Cookie Documentation](https://github.com/js-cookie/js-cookie/tree/latest#readme)
 */

File: web-greeter/bridge/Greeter.py
# -*- coding: utf-8 -*-
#
#  Greeter.py
#
#  Copyright © 2017 Antergos
#  Copyright © 2021 JezerM
#
#  This file is part of Web Greeter.
#
#  Web Greeter is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License, or
#  (at your option) any later version.
#
#  Web Greeter is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  The following additional terms are in effect as per Section 7 of the license:
#
#  The preservation of all legal notices and author attributions in
#  the material or in the Appropriate Legal Notices displayed
#  by works containing it is required.
#
#  You should have received a copy of the GNU General Public License
#  along with Web Greeter; If not, see <http://www.gnu.org/licenses/>.

# Standard Lib
import subprocess
import threading

# 3rd-Party Libs
import gi
gi.require_version('LightDM', '1')
from gi.repository import LightDM

from browser.bridge import Bridge, BridgeObject
from PyQt5.QtCore import QVariant, QTimer

from config import web_greeter_config

# This Application
from . import (
    language_to_dict,
    layout_to_dict,
    session_to_dict,
    user_to_dict,
    battery_to_dict,
    logger
)

# import utils.battery as battery

LightDMGreeter = LightDM.Greeter()
LightDMUsers = LightDM.UserList()


def changeBrightness(self, method: str, quantity: int):
    if self._config["features"]["backlight"]["enabled"] != True:
        return
    try:
        steps = self._config["features"]["backlight"]["steps"]
        child = subprocess.run(["xbacklight", method, str(quantity), "-steps", str(steps)])
        if child.returncode == 1:
            raise ChildProcessError("xbacklight returned 1")
    except Exception as err:
        logger.error("Brightness: {}".format(err))
    else:
        self.brightness_update.emit()


def getBrightness(self):
    if self._config["features"]["backlight"]["enabled"] != True:
        return -1
    try:
        level = subprocess.run(["xbacklight", "-get"], stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE, text=True, check=True)
        return int(level.stdout)
    except Exception as err:
        logger.error("Brightness: {}".format(err))
        return -1

class Greeter(BridgeObject):

    # LightDM.Greeter Signals
    authentication_complete = Bridge.signal()
    autologin_timer_expired = Bridge.signal()
    idle = Bridge.signal()
    reset = Bridge.signal()
    show_message = Bridge.signal(str, LightDM.MessageType, arguments=('text', 'type'))
    show_prompt = Bridge.signal(str, LightDM.PromptType, arguments=('text', 'type'))

    brightness_update = Bridge.signal()
    battery_update = Bridge.signal()

    noop_signal = Bridge.signal()
    property_changed = Bridge.signal()

    _battery = None

    def __init__(self, *args, **kwargs):
        super().__init__(name='LightDMGreeter', *args, **kwargs)

        self._config = web_greeter_config["config"]
        self._shared_data_directory = ''
        self._themes_directory = web_greeter_config["app"]["theme_dir"]

        # if self._config.features.battery == True:
            # self._battery = battery.Battery()

        LightDMGreeter.connect_to_daemon_sync()

        self._connect_signals()
        self._determine_shared_data_directory_path()

    def _determine_shared_data_directory_path(self):
        user = LightDMUsers.get_users()[0]
        user_data_dir = LightDMGreeter.ensure_shared_data_dir_sync(user.get_name())
        self._shared_data_directory = user_data_dir.rpartition('/')[0]

    def _connect_signals(self):
        LightDMGreeter.connect(
            'authentication-complete',
            lambda greeter: self._emit_signal(self.authentication_complete)
        )
        LightDMGreeter.connect(
            'autologin-timer-expired',
            lambda greeter: self._emit_signal(self.autologin_timer_expired)
        )

        LightDMGreeter.connect('idle', lambda greeter: self._emit_signal(self.idle))
        LightDMGreeter.connect('reset', lambda greeter: self._emit_signal(self.reset))

        LightDMGreeter.connect(
            'show-message',
            lambda greeter, msg, mtype: self._emit_signal(self.show_message, msg, mtype)
        )
        LightDMGreeter.connect(
            'show-prompt',
            lambda greeter, msg, mtype: self._emit_signal(self.show_prompt, msg, mtype)
        )

        if self._battery:
            self._battery.connect(lambda: self.battery_update.emit())

    def _emit_signal(self, _signal, *args):
        self.property_changed.emit()
        QTimer().singleShot(300, lambda: _signal.emit(*args))

    @Bridge.prop(str, notify=property_changed)
    def authentication_user(self):
        return LightDMGreeter.get_authentication_user() or ''

    @Bridge.prop(bool, notify=noop_signal)
    def autologin_guest(self):
        return LightDMGreeter.get_autologin_guest_hint()

    @Bridge.prop(int, notify=noop_signal)
    def autologin_timeout(self):
        return LightDMGreeter.get_autologin_timeout_hint()

    @Bridge.prop(str, notify=noop_signal)
    def autologin_user(self):
        return LightDMGreeter.get_autologin_user_hint()

    @Bridge.prop(QVariant, notify=battery_update)
    def batteryData(self):
        return battery_to_dict(self._battery)

    @Bridge.prop(int, notify=brightness_update)
    def brightness(self):
        return getBrightness(self)

    @brightness.setter
    def brightness(self, quantity):
        thread = threading.Thread(target=changeBrightness,
                                  args=(self, "-set", quantity))
        thread.start()

    @Bridge.prop(bool, notify=noop_signal)
    def can_hibernate(self):
        return LightDM.get_can_hibernate()

    @Bridge.prop(bool, notify=noop_signal)
    def can_restart(self):
        return LightDM.get_can_restart()

    @Bridge.prop(bool, notify=noop_signal)
    def can_shutdown(self):
        return LightDM.get_can_shutdown()

    @Bridge.prop(bool, notify=noop_signal)
    def can_suspend(self):
        return LightDM.get_can_suspend()

    @Bridge.prop(bool, notify=noop_signal)
    def can_access_brightness(self):
        return self._config["features"]["backlight"]["enabled"]

    @Bridge.prop(bool, notify=noop_signal)
    def can_access_battery(self):
        return self._config["features"]["battery"]

    @Bridge.prop(str, notify=noop_signal)
    def default_session(self):
        return LightDMGreeter.get_default_session_hint()

    @Bridge.prop(bool, notify=noop_signal)
    def has_guest_account(self):
        return LightDMGreeter.get_has_guest_account_hint()

    @Bridge.prop(bool, notify=noop_signal)
    def hide_users_hint(self):
        return LightDMGreeter.get_hide_users_hint()

    @Bridge.prop(str, notify=noop_signal)
    def hostname(self):
        return LightDM.get_hostname()

    @Bridge.prop(bool, notify=property_changed)
    def in_authentication(self):
        return LightDMGreeter.get_in_authentication()

    @Bridge.prop(bool, notify=property_changed)
    def is_authenticated(self):
        return LightDMGreeter.get_is_authenticated()

    @Bridge.prop(QVariant, notify=property_changed)
    def language(self):
        return language_to_dict(LightDM.get_language())

    @Bridge.prop(QVariant, notify=noop_signal)
    def languages(self):
        return [language_to_dict(lang) for lang in LightDM.get_languages()]

    @Bridge.prop(QVariant, notify=noop_signal)
    def layout(self):
        return layout_to_dict(LightDM.get_layout())

    @layout.setter
    def layout(self, layout):
        if type(layout) != dict:
            return False
        lay = dict(
            name = layout.get("name") or "",
            description = layout.get("description") or "",
            short_description = layout.get("short_description") or ""
        )
        return LightDM.set_layout(LightDM.Layout(**lay))

    @Bridge.prop(QVariant, notify=noop_signal)
    def layouts(self):
        return [layout_to_dict(layout) for layout in LightDM.get_layouts()]

    @Bridge.prop(bool, notify=noop_signal)
    def lock_hint(self):
        return LightDMGreeter.get_lock_hint()

    @Bridge.prop(QVariant, notify=property_changed)
    def remote_sessions(self):
        return [session_to_dict(session) for session in LightDM.get_remote_sessions()]

    @Bridge.prop(bool, notify=noop_signal)
    def select_guest_hint(self):
        return LightDMGreeter.get_select_guest_hint()

    @Bridge.prop(str, notify=noop_signal)
    def select_user_hint(self):
        return LightDMGreeter.get_select_user_hint() or ''

    @Bridge.prop(QVariant, notify=noop_signal)
    def sessions(self):
        return [session_to_dict(session) for session in LightDM.get_sessions()]

    @Bridge.prop(str, notify=noop_signal)
    def shared_data_directory(self):
        return self._shared_data_directory

    @Bridge.prop(bool, notify=noop_signal)
    def show_manual_login_hint(self):
        return LightDMGreeter.get_show_manual_login_hint()

    @Bridge.prop(bool, notify=noop_signal)
    def show_remote_login_hint(self):
        return LightDMGreeter.get_show_remote_login_hint()

    @Bridge.prop(str, notify=noop_signal)
    def themes_directory(self):
        return self._themes_directory

    @Bridge.prop(QVariant, notify=noop_signal)
    def users(self):
        return [user_to_dict(user) for user in LightDMUsers.get_users()]

    @Bridge.method(str)
    def authenticate(self, username):
        LightDMGreeter.authenticate(username)
        self.property_changed.emit()

    @Bridge.method()
    def authenticate_as_guest(self):
        LightDMGreeter.authenticate_as_guest()
        self.property_changed.emit()

    @Bridge.method(int)
    def brightnessSet(self, quantity):
        thread = threading.Thread(target=changeBrightness,
                                  args=(self, "-set", quantity))
        thread.start()

    @Bridge.method(int)
    def brightnessIncrease(self, quantity):
        thread = threading.Thread(target=changeBrightness,
                                  args=(self, "-inc", quantity))
        thread.start()

    @Bridge.method(int)
    def brightnessDecrease(self, quantity):
        thread = threading.Thread(target=changeBrightness,
                                  args=(self, "-dec", quantity))
        thread.start()

    @Bridge.method()
    def cancel_authentication(self):
        LightDMGreeter.cancel_authentication()
        self.property_changed.emit()

    @Bridge.method()
    def cancel_autologin(self):
        LightDMGreeter.cancel_autologin()
        self.property_changed.emit()

    @Bridge.method(result=bool)
    def hibernate(self):
        return LightDM.hibernate()

    @Bridge.method(str)
    def respond(self, response):
        LightDMGreeter.respond(response)
        self.property_changed.emit()

    @Bridge.method(result=bool)
    def restart(self):
        return LightDM.restart()

    @Bridge.method(str)
    def set_language(self, lang):
        if self.is_authenticated:
            LightDMGreeter.set_language(lang)
            self.property_changed.emit()

    @Bridge.method(result=bool)
    def shutdown(self):
        return LightDM.shutdown()

    @Bridge.method(str, result=bool)
    def start_session(self, session):
        if not session.strip():
            return
        return LightDMGreeter.start_session_sync(session)

    @Bridge.method(result=bool)
    def suspend(self):
        return LightDM.suspend()
